### DISCUSSION TITLE
Cannot figure out how to get started with Syphon in Ultragrid

### DISCUSSION THREAD

**Discussion by ygreq** [CESNET/UltraGrid#143] [Q&A]
Hi there!

I managed to set a duplex communication between a pc and a mac for now on a local network. On the pc side, I managed to work with Spout, but on the MacOS side, I cannot manage to get going with Syphon. 

On MacOS, this works alright using the integrated webcam. Some frames are lost, probably from the codec. 
`./uv --capture-filter preview -t avfoundation:device=0:mode=0:fps=30 -c libavcodec:codec=MJPEG -d multiplier:gl#preview --control-port 8888 [IP_ADDRESS] --param errors-fatal`

But this does not:
`./uv --capture-filter preview -t syphon:name=fromMax:app=Max:override_fps=60 -c libavcodec:codec=H.264 -d gl:syphon=toMax --control-port 8888 [IP_ADDRESS] --param errors-fatal
`
It gives me this error: 
_Unable to set socket buffer size: No buffer space available
Cannot run display when another mainloop registered!_

BTW, I am running a syphon server and a syphon receiver on Cycling '74 Max 8. 
And I still did not figure out what "app: Syphon server application name" really mean? Am I right to type Max considering the syphon server is on Max? Can I type anything?

Thank you so much for all your help!
ygreq

**Comment by MartinPulec** [comment]
Hi @ygreq,

> Cannot run display when another mainloop registered!

this is a technical problem with macOS. Some, mainly GUI related tasks, need to be run from within the main event loop running in main thread. Such tasks UltraGrid deploys to main thread. But because Syphon capture needs to be run there, the display task then cannot. Easiest workaround would be to use separate UltraGrid processes for capture and display. In theory, the GL display doesn't need to run in the main thread but the window will then be a bit crippled (won't accept keyboard input and other window events like fullscreen or close). If you wish, I can add an option for that, but I think it is suboptimal.

> BTW, I am running a syphon server and a syphon receiver on Cycling '74 Max 8.
>And I still did not figure out what "app: Syphon server application name" really mean? Am I right to type Max considering the syphon server is on Max? Can I type anything?

I am not completely acquainted with Syphon nomenclature but AFAIK the server has 2 properties _name_ and _application name_. Perhaps with different name the application may differentiate multiple servers while the _application name_ remains the same? Anyways, it should be needed only to set only one of those if the value is unique.

Anyways, the options `--capture-filter preview` and `-d multiplier:XXX#preview` and `--control-port 8888` are needless if you are not running UG from GUI (well, control-port could be used, in theory).

**Reply by ygreq** [reply]
I abandoned this thread as I did not understand the solution you proposed in order to make Syphon run.
And I still did not find a solution for this :(

I send this:

.`/uv -t syphon:name=fromMax_Mac:app=uv:override_fps=60 -c libavcodec:codec=H.264 -d gl:syphon=toMax_Mac 192.168.1.104`

This is what I get:
```
MasterPortDestroyer is armed
UltraGrid 1.6 (tags/v1.6 rev 1618720e built Jan  5 2021 14:55:04)

Display device   : gl
Capture device   : syphon
Audio capture    : none
Audio playback   : none
MTU              : 1500 B
Video compression: libavcodec:codec=H.264
Audio codec      : PCM
Network protocol : UltraGrid RTP
Audio FEC        : none
Video FEC        : none

Created new RTP session with SSRC 0x1f66357b.
GL setup: fullscreen: OFF, deinterlace: OFF
Display initialized-gl
Video capture initialized-syphon
Created new RTP session with SSRC 0x52121b1c.
Unable to set socket buffer size: No buffer space available
Cannot run display when another mainloop registered!

Exit
MasterPortDestroyer triggered
```

I also tried running this: 
`./uv -t syphon:name=fromMax_Mac:app=uv:override_fps=60 -c libavcodec:codec=H.264 -d gl 192.168.1.104`
but still same error

**Reply by MartinPulec** [reply]
The problem is that you cannot run the _capture_ (`-t` parameter) and _playback_ (the `-d` parameter) from one process here if using Syphon. What about running 2 processes - one for sender, the other for receiver?

    ./uv -t syphon:name=fromMax_Mac:app=uv:override_fps=60 -c libavcodec:codec=H.264 192.168.1.104
    ./uv -d gl 192.168.1.104

This would work and should behave the same as your command.

**Reply by ygreq** [reply]
I see. I didn't know you could run them separately like this. 

But how do you run them both? If I start one process, I am not allowed to start the second until I stop the first. I tried with `./uv -t syphon:name=Jitter:app=Max:override_fps=60 -c libavcodec:codec=H.264 192.168.1.104 && -d gl 192.168.1.104`, but it gives me this and no video sent or received: 

```
MasterPortDestroyer is armed
UltraGrid 1.6 (tags/v1.6 rev 1618720e built Jan  5 2021 14:55:04)

Display device   : none
Capture device   : syphon
Audio capture    : none
Audio playback   : none
MTU              : 1500 B
Video compression: libavcodec:codec=H.264
Audio codec      : PCM
Network protocol : UltraGrid RTP
Audio FEC        : none
Video FEC        : none

Created new RTP session with SSRC 0x186e5aa2.
Display initialized-none
Video capture initialized-syphon
Created new RTP session with SSRC 0x0146d955.
Unable to set socket buffer size: No buffer space available
[Syphon capture] Using server - app: Max name: Jitter
[lavc] Using codec: H.264, encoder: libx264
[lavc] Setting CRF to 22.00.
[lavc] Warning: Codec doesn't support slice-based multithreading.
[lavc] Setting preset to ultrafast.
[libx264 @ 0x7fa12606f200] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
[libx264 @ 0x7fa12606f200] profile Constrained Baseline, level 2.1, 4:2:0, 8-bit
[lavc] Selected pixfmt: yuv420p
[Syphon capture] 277 frames in 5.01507 seconds = 55.2335 FPS
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] 293 frames in 5.00015 seconds = 58.5983 FPS
[Syphon capture] 301 frames in 5.02443 seconds = 59.9072 FPS
[Syphon capture] 301 frames in 5.00822 seconds = 60.1012 FPS
[Syphon capture] 300 frames in 5.00185 seconds = 59.9777 FPS
[Syphon capture] 299 frames in 5.00458 seconds = 59.7453 FPS
[Syphon capture] 202 frames in 5.02187 seconds = 40.224 FPS
[Syphon capture] 188 frames in 5.00649 seconds = 37.5513 FPS
[Syphon capture] 198 frames in 5.03781 seconds = 39.3028 FPS
[Syphon capture] 200 frames in 5.00414 seconds = 39.9669 FPS
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] 146 frames in 5.0071 seconds = 29.1586 FPS
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] 116 frames in 5.09777 seconds = 22.755 FPS
[Syphon capture] 144 frames in 5.0066 seconds = 28.762 FPS
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
[Syphon capture] Skipping frame.
```

**Reply by MartinPulec** [reply]
Is the command you've written really the one you are using? It doesn't make any sense to me - using the operator `&&` causes sequential run of the processes, not concurrent. You could probably use `&` instead but I would not recommend that. I'd use either 2 terminals/tabs or a terminal multiplexer if not possible (eg. [tmux](https://github.com/tmux/tmux/wiki)). Also there is a typo in your second command - the actual UG executable name is missing.

As for the `Skipping frame` message - it looks like that this is a performance issue - I could increase buffer length but I am not sure if it helps, it only allows to cover the performance peaks but not a power lack. But you can check, how much does the process take of CPU time, if not much, it may be a way.

**Reply by MartinPulec** [reply]
Update also on that - you may try to use syphon option `queue_size`, eg. `-t syphon:queue_size=3` to improve performance. If it really helps, I'd be glad if you could let me know - I may adjust the default value then.

**Reply by ygreq** [reply]
Thank you, Martin! I will test asap and let you know.

**Reply by ygreq** [reply]
Ok. I finally got to test on the MacBook Air. 

I tried sending the webcam feedback with a 320x240 resolution. The code I used is: 
`./uv -t syphon:name=Jitter:app=Max:override_fps=60 -c libavcodec:codec=H.264 192.168.1.104` 
and separately
`./uv -d gl 192.168.1.104`

The image I received from the other laptop (powerful PC) is top notch. The image I sent even though small, was skipping frames 

I also tested `./uv -t syphon:queue_size=3:name=Jitter:app=Max:override_fps=60 -c libavcodec:codec=H.264 192.168.1.104` I got an error saying Syphon: Unknown argument - queue_size=3. Unable to start video capture device syphon

So as a conclusion it is definitely related to the fact that the MacBook Air is not powerful enough. 

In the meantime I got another powerful Windows laptop so I will be using that one for further tests. If you want me to do further test, let me know. Otherwise for now I will stop using Ultragrid on MacOS. 

Thank you so much for all your help!
